src/v/s3/client.cc

```diff
@@ -499,4 +505,55 @@ client::delete_object(const bucket_name& bucket, const object_key& key) {
       });
 }
 
+client_pool::client_pool(
+  size_t size, configuration conf, client_pool_overdraft_policy policy)
+  : _size(size)
+  , _config(std::move(conf))
+  , _policy(policy) {
+    init();
+}
+
+ss::future<> client_pool::stop() {
+    _as.request_abort();
+    return _gate.close();
+}
+
+/// \brief Acquire http client from the pool.
+///
+/// \note it's guaranteed that the client can only be acquired once
+///       before it gets released (release happens implicitly, when
+///       the lifetime of the pointer ends).
+/// \return client pointer (via future that can wait if all clients
+///         are in use)
+ss::future<client_pool::client_lease> client_pool::acquire() {
+    gate_guard guard(_gate);
+    if (_pool.empty()) {
+        if (_policy == client_pool_overdraft_policy::wait_if_empty) {
+            co_await _cvar.wait([this] { return !_pool.empty(); });
```

Comment:
question, will this keep the `guard` alive given that it's not referenced anywhere? looks like a violation to me no?

Reply:
I thkin that this  is also an instance of gcc coroutine temporary bug.